vignettes/mcstate2.Rmd

```diff
@@ -0,0 +1,142 @@
+
+Draw samples from a model using Markov Chain Monte Carlo methods.  To do this you will need:
+
+* A **model**: this is an `mcstate_model()` object and minimally knows about the names of its parameter vector (which is an unstructured real-valued vector) and can compute a log probability density.  It may also be able to compute a gradient with of this log density, or sample directly from parameter space (e.g., if it represents a prior distribution).
+* A **sampler**: this is some method of drawing samples from the models distribution in a sequence.  We define several different sampler types, with the simplest one being `mcstate_sampler_random_walk()`, which implements a simple Metropolis algorithm random walk.
+* A **runner**: this controls how the chains will be run (e.g., one after another or in parallel).
+
+The system is designed to be composable; you can work in a Bayesian way by defining a model representing a likelihood and another model representing a prior and then pick a sampler based on the capabilities of the model, and pick a runner based on the capabilities your computer.
+
+The `mcstate_model()` interface is designed to be very flexible but not user-friendly.  We expect to write a higher-level interface to help work with this, and describe how to write wrappers for models implemented in other packages (so you might write a model in [dust](https://mrc-ide.github.io/dust) or [odin](https://mrc-ide.github.io/odin) and an adaptor would make it easy to work with the tools provided by `mcstate2` to start making inferences with your model).
+
+# An example
+
+Before starting the example, it's worth noting that there are far better tools out there to model this sort of thing (stan, bugs, jags, R itself - really anything).  The aim of this section is to derive a simple model that may feel familiar.  The strength of the package is when performing inference with custom models that can't be expressed in these high level interfaces.
+
+```{r, include = FALSE}
+set.seed(1)
+data <- local({
+  n <- 60
+  weight <- rnorm(n, 50, 6.5)
+  height <- 114 + weight * 0.9 + rnorm(60, sd = 3)
+  data.frame(height, weight)
+})
+```
+
+
+```{r}
+head(data)
+plot(height ~ weight, data)
+```
+
+A simple likelihood, following the model formulation in "Statistical Rethinking" chapter 3; height is modelled as normally distributed departures from a linear relationship with weight.
+
+```{r}
+likelihood <- mcstate_model(
+  list(
+    parameters = c("a", "b", "sigma"),
+    density = function(x) {
+      a <- x[[1]]
+      b <- x[[2]]
+      sigma <- x[[3]]
+      mu <- a + b * data$weight
+      sum(dnorm(data$height, mu, sigma, log = TRUE))
+    }))
+```
+
+The prior we'll make much nicer to work with in future, but here we construct the density by hand as a sum of normally distributed priors on `a` and `b`, and a weak uniform prior on `sigma`.  We provide a `direct_sample` function here so that we can draw samples from the prior distribution directly.
```

Comment:
```suggestion
The prior we'll make much nicer to work with in the future, but here we construct the density by hand as a sum of normally distributed priors on `a` and `b`, and a weak uniform prior on `sigma`.  We provide a `direct_sample` function here so that we can draw samples from the prior distribution directly.
```